compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala

```diff
@@ -29,20 +21,64 @@ object NullOpsDecorator {
       self.isDirectRef(defn.JavaNullAlias)
     }
 
+    /** Normalizes unions so that all `Null`s (or aliases to `Null`) appear to the right of
+     *  all other types.
+     *  e.g. `Null | (T1 | Null) | T2` => `T1 | T2 | Null`
+     *  e.g. `JavaNull | (T1 | Null) | Null` => `T1 | Null | JavaNull`
+     *
+     *  1. If self is not a union type, the result is not a union type.
+     *  2. If self is a union type,
+     *    (1) If it only contains `Null` (or `JavaNull`) types, the result is still a union
+              only contains `Null` types.
+     *    (2) If it contains multiple Null types, at most 2 different Null types
+     *        will appear at the right of result.
+     *    (3) If it contains both `Null` and `JavaNull`, the result is `(tp | Null) | JavaNull`.
+     *    (4) If it onlt contains `Null` or `JavaNull`, the result is `tp | Null` or `tp | JavaNull`.
+     */
+     def normNullableUnion(implicit ctx: Context): Type = {
+      var isUnion = false
+      var hasNull = false
+      var hasJavaNull = false
+      def strip(tp: Type): Type = tp match {
+        case tp @ OrType(lhs, rhs) =>
+          isUnion = true
+          val llhs = strip(lhs)
+          val rrhs = strip(rhs)
+          if (rrhs.isNullType) llhs
+          else if (llhs.isNullType) rrhs
+          else tp.derivedOrType(llhs, rrhs)
+        case _ =>
+          if (tp.isNullType) {
+            if (tp.isJavaNullType) hasJavaNull = true
+            else hasNull = true
+          }
+          tp
+      }
+      val tp = strip(self)
+      if (!isUnion) self
+      else {
+        val tp1 = if (hasNull) OrType(tp, defn.NullType) else tp
```

Comment:
This will need to change to only return the `JavaNull` if both Null and JavaNull are present.